### PR TITLE
WiP refactor: Manage SpoonClassNotFoundException in getActualClass

### PR DESCRIPTION
**Pull request by surli**
This PR is an alternative of #1946 for fixing #1833

### CODE REVIEW

File: src/main/java/spoon/support/SpoonInternalException.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ *
+ * The fact that you are presently reading this means that you have had
+ * knowledge of the CeCILL-C license and that you accept its terms.
+ */
+package spoon.support;
+
+/** Exception to be managed internally
+ * All external Spoon methods should only return {@link spoon.SpoonException}
+ */
+public class SpoonInternalException extends Exception {
```
[the remark this answers is not in the document]
Comment (surli) [reply src/main/java/spoon/support/SpoonInternalException.java]: I don't understand what you meant by:
> We would differentiate between SpoonException (for end users) and SpoonInternalException (for us to catch).

if we want to catch `SpoonInternalException` they should not extend `RuntimeException`.

**Comment by monperrus** [comment]
We want to catch it but maybe not in the caller directly. To avoid polluting method signatures, we
make it RuntimeException. This is the standard practice used in Spoon for years.

**Comment by surli** [comment]
> To avoid polluting method signatures

Here the only signature change is on `findClass`, which should not be overridden by the clients.

> This is the standard practice used in Spoon for years.

Yeah and that's why we now have to refactor `getActualClass`: if you make this internal exception extended runtime it becomes useless from my point of view. The purpose of this class is actually to be seen in the method signature in order to be catched and managed properly.

**Comment by monperrus** [comment]
The debate between checked and unchecked is an old one in the Java community.

I'd be happy to convince you that the choice of only using unchecked is the best one. That's why
Spring, Spoon and other top Java projects do so. We can discuss this over a beer :-)

It seems to me that we can proceed with fixing the issue without opening Pandora's box of
checked/unchecked.

**Comment by surli** [comment]
> It seems to me that we can proceed with fixing the issue without opening Pandora's box of checked/unchecked.

I understand that you don't want to open this pandora box for external API, but I don't understand what bothers you with changing it for a very small part of the internal API.

My goal here is to ease the future maintaining: I really believe that if you keep it unchecked, the entire PR is completely useless because we would have to do the same job in few monthes when me or another maintainer would use this method in Spoon and don't see that he should manage a potential exception.

**Comment by monperrus** [comment]
The arguments against checked exceptions hold for both the internal and external API. See:

- https://www.javacodegeeks.com/2012/03/why-should-you-use-unchecked-exceptions.html
- https://docs.oracle.com/javase/tutorial/essential/exceptions/runtime.html

I agree with you, making future contributors aware is important, but IMHO checked exceptions is not the most appropriate solution.

**Comment by surli** [comment]
> Here's the bottom line guideline: If a client can reasonably be expected to recover from an exception, make it a checked exception. 

https://docs.oracle.com/javase/tutorial/essential/exceptions/runtime.html

And it's actually the case for this noClasspath case.

**Comment by monperrus** [comment]
so you really want to officially start the debate :-)?

and at the same occasion, to start a revolution by questioning one of the very old design rule of Spoon :-D?

File: src/main/java/spoon/support/SpoonInternalException.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (monperrus) [reply src/main/java/spoon/support/SpoonInternalException.java]: per our discussion "SpoonFindClassException extends SpoonClassNotFoundException"

**Pull request closed without merging**